doc/contract-update.md

```diff
@@ -0,0 +1,60 @@
+# Upgrading and Migrating contracts
+
+We consider two approaches:  *Upgrading* existing contracts to a new version but retaining
+state data as-is, and; *migrating* users to a completely new set of contracts.
+
+We would only migrate in the event of a key loss, which prevents upgrading contracts.
+
+## Upgrade contracts approach
+
+Our contracts are already mostly divided into contracts that contain data and contracts that
+contain logic. This is fortuitous for the pattern we are going to use.
+
+`State.sol` contains most of the state, with some in `Shield.sol`.
+This division could be made complete by moving the `proposeBlock` function out of `State.sol`
+and moving the `Shield.sol` data into `State.sol` (there may be a few other minor adjustments
+TBC). `State.sol` will then only contain storage data and getters/setters for the same.
+It will also need modifying so that the addresses of the logic contracts, which it allows to
+set state, can be updated by a registered contract containing an `upgrade` function (see later).
+
+After that is done, we will never upgrade the `State.sol` contract but will upgrade the
+stateless logic contracts. The logic contracts call `State.sol` and thus we avoid the use of
+`delegateCall`. If, in the future, `State.sol` becomes insufficient for our needs, another
+data contract can be deployed alongside, which adds the new storage variables needed.
```

Comment:
There might be instances where we would want to replace the current state variable other than adding new variables. For example, with the new PoS that David is implementing we replace `BondAccounts` with `stakeAccounts`. Thankfully they are of the same data types and hold similar data. So if we were to upgrade to the new PoS approach, then we would simply continue using this variable. But say if we were to make changes to PoS such that the bondAccounts variable's data type and data aren't suitable anymore, then we want to stop using the old variable and start using a new variable. Pointing the logic contracts to the new variable might suffice but sometimes it would also need data migration from old to new variable. In this example, the proposers can deregister and register with stakes again so they are reflected in the new state. A similar migration might be sufficient for proposer or challenger related variables. Even for users, the `shield.sol` only stores `withdraw,  advancedFeeWithdrawals, advancedWithdrawals, feebook` variables. So a migration by users for `withdraw,  advancedFeeWithdrawals, advancedWithdrawals` variables would mean pausing new advance withdrawals and for `feebook` would mean pausing deposits, transfers and withdraws , then waiting 7 days to confirm any remaining transactions. 

So we might want to consider adding a pause contract function?
This seems important even its not for migrating variables.

Reply:
Yes, this makes sense. Using it would be a very bad user experience but we might one day find that it's our only recourse.